# erreur

Un Outil [Sumor Cloud](https://sumor.cloud).  
[Plus de Documentation](https://sumor.cloud/error)  
Une bibliothèque de gestion d'erreur qui prend en charge le multilangage, les codes d'erreur prédéfinis et le passage de données.

[![CI](https://github.com/sumor-cloud/error/actions/workflows/ci.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/ci.yml)  
[![Test](https://github.com/sumor-cloud/error/actions/workflows/ut.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/ut.yml)  
[![Coverage](https://github.com/sumor-cloud/error/actions/workflows/coverage.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/coverage.yml)  
[![Audit](https://github.com/sumor-cloud/error/actions/workflows/audit.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/audit.yml)

## Installation

```bash
npm i @sumor/error --save
```

## Prérequis

### Version Node.JS

Nécessite la version Node.JS 18.x ou supérieure

### Requiert un module ES de Node.JS

Comme ce package est écrit en module ES, veuillez modifier le code suivant dans votre fichier `package.json` :

```json
{
  "type": "module"
}
```

## Utilisation

### Définir une Erreur

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    USER_NOT_FOUND: 'Utilisateur non trouvé',
    USER_EXISTED: 'Utilisateur {name} existe'
  }
})

throw new MyError('USER_NOT_FOUND')
// sortie: Error: Utilisateur non trouvé

throw new MyError('USER_EXISTED', { name: 'Alice' })
// sortie: Error: L'utilisateur Alice existe
```

### Multilangage

```js
import defineError from '@sumor/error'

const MyError = defineError({
  language: 'en', // langue par défaut
  code: {
    USER_NOT_FOUND: 'Utilisateur non trouvé',
    USER_EXISTED: 'Utilisateur {name} existe'
  },
  i18n: {
    zh: {
      USER_NOT_FOUND: 'Utilisateur non trouvé',
      USER_EXISTED: 'Utilisateur {name} existe'
    }
  }
})

const error = new MyError('USER_EXISTED', { name: 'Alice' })
error.language = 'en' // changer la langue de l'erreur
console.log(error)
// sortie: Error: L'utilisateur Alice existe

error.language = 'zh' // changer la langue de l'erreur
console.log(error)
// sortie: Error: L'utilisateur Alice existe
```

### Convertir une Erreur en JSON

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    USER_NOT_FOUND: 'Utilisateur non trouvé',
    USER_EXISTED: 'Utilisateur {name} existe'
  }
})

const error = new MyError('USER_EXISTED', { name: 'Alice' })
console.log(error.json())
// sortie: {"code":"USER_EXISTED","message":"L'utilisateur Alice existe"}
```

### Erreur Sous-jacente

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    FIELD_VERIFY_FAILED: 'Échec de la vérification du champ',
    FIELD_CANNOT_EMPTY: 'Le champ {name} ne peut pas être vide',
    FIELD_TOO_LONG: 'Le champ {name} est trop long'
  },
  i18n: {
    zh: {
      FIELD_VERIFY_FAILED: 'Échec de la vérification du champ',
      FIELD_CANNOT_EMPTY: 'Le champ {name} ne peut pas être vide',
      FIELD_TOO_LONG: 'Le champ {name} est trop long'
    }
  }
})

const error = new MyError('FIELD_VERIFY_FAILED', {}, [
  new MyError('FIELD_CANNOT_EMPTY', { name: 'nom d'utilisateur' }),
  new MyError('FIELD_TOO_LONG', { name: 'mot de passe' })
])

console.log(error.json())
/*
output:
{
  "code":"FIELD_VERIFY_FAILED",
  "message":"Échec de la vérification du champ",
  "errors":[
    {
      "code":"FIELD_CANNOT_EMPTY",
      "message":"Le champ nom d'utilisateur ne peut pas être vide"
    },{
      "code":"FIELD_TOO_LONG",
      "message":"Le champ mot de passe est trop long"
    }
  ]
}
*/

error.language = 'zh'
console.log(error.json())
/*
output:
{
  "code":"FIELD_VERIFY_FAILED",
  "message":"Échec de la vérification du champ",
  "errors":[
    {
      "code":"FIELD_CANNOT_EMPTY",
      "message":"Le champ nom d'utilisateur ne peut pas être vide"
    },{
      "code":"FIELD_TOO_LONG",
      "message":"Le champ mot de passe est trop long"
    }
  ]
}
*/
```

### Combinaison d'Erreurs Standard

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    FIELD_VERIFY_FAILED: 'Échec de la vérification du champ',
    FIELD_CANNOT_EMPTY: 'Le champ {name} ne peut pas être vide',
    FIELD_TOO_LONG: 'Le champ {name} est trop long'
  }
})

const error = new MyError('FIELD_VERIFY_FAILED', {}, [
  new MyError('FIELD_CANNOT_EMPTY', { name: 'nom d'utilisateur' }),
  new MyError('FIELD_TOO_LONG', { name: 'mot de passe' }),
  new Error('Erreur Inconnue')
])

console.log(error.json())
/*
output:
{
  "code":"FIELD_VERIFY_FAILED",
  "message":"Échec de la vérification du champ",
  "errors":[
    {
      "code":"FIELD_CANNOT_EMPTY",
      "message":"Le champ nom d'utilisateur ne peut pas être vide"
    },{
      "code":"FIELD_TOO_LONG",
      "message":"Le champ mot de passe est trop long"
    },{
      "code":"UNKNOWN_ERROR",
      "message":"Erreur Inconnue"
    }
  ]
}
 */
```
